### PR TITLE
Ble updates

### PR DESCRIPTION
**Pull request by mikkeldamsgaard**
Added support for MTU, fixing some issues creeping up when there is high throughput and fixed an issue with a malloc error that is not really a malloc error

### CODE REVIEW

File: tests/ctest/external_messaging_test.cc [hunk 1]
```diff
@@ -37,6 +37,7 @@ class MessageHandler : public ExternalSystemMessageHandler {
   explicit MessageHandler(VM* vm) : ExternalSystemMessageHandler(vm) { }
   virtual void on_message(int sender, int type, void* data, int length) override;
 
+
```
Comment (kasperl) [inline tests/ctest/external_messaging_test.cc:40]: Drop the whitespace only changes? (this entire file)

File: tests/ctest/external_messaging_input.toit [hunk 2]
```diff
@@ -21,6 +21,17 @@ main:
   test handler (ByteArray 3197: it)
   test handler (ByteArray 31971: it)
 
+  // send 2k buffer to the external process, and receive 100 bytes.
```
Comment (kasperl) [inline tests/ctest/external_messaging_input.toit:24]: This appears unrelated to the BLE updates?

File: src/resources/ble_esp32.cc [hunk 3]
```diff
@@ -477,7 +481,11 @@ static Object* object_to_mbuf(Process* process, Object* object, os_mbuf** result
     if (!object->byte_content(process->program(), &bytes, STRINGS_OR_BYTE_ARRAYS)) WRONG_TYPE;
     if (bytes.length() > 0) {
       os_mbuf* mbuf = ble_hs_mbuf_from_flat(bytes.address(), bytes.length());
-      if (!mbuf) MALLOC_FAILED;
+      // A null response is not an allocation error, as the mbufs are allocated on boot based on configuration settings.
+      // Therefore a QC will do little to help the situation and the VM will crash
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:485]:
```suggestion
      // Therefore a GC will do little to help the situation and the VM will think it is out of memory.
```

File: lib/ble.toit [hunk 4]
```diff
@@ -588,13 +626,37 @@ ble_add_server_service_ resource_group_ uuid:
...
 ble_get_characteristics_value_ gatt:
   #primitive.ble.get_characteristics_value
+
+ble_get_mtu_ gatt:
+  #primitive.ble.get_att_mtu
+
+ble_run_with_quota_backoff_ [block]:
+  u := Time.monotonic_us
+  while true:
+    e := catch:
```
Comment (kasperl) [inline lib/ble.toit:655]:
```suggestion
    catch --unwind=(: it != "QUOTA_EXCEEDED"): return block.call
    sleep --ms=10
    if Time.monotonic_us - start > 2_000_000: throw DEADLINE_EXCEEDED_ERROR
```

File: lib/ble.toit [hunk 5]
```diff
@@ -588,13 +626,37 @@ ble_add_server_service_ resource_group_ uuid:
...
   #primitive.ble.notify_characteristics_value
 
 ble_get_characteristics_value_ gatt:
   #primitive.ble.get_characteristics_value
+
+ble_get_mtu_ gatt:
+  #primitive.ble.get_att_mtu
+
+ble_run_with_quota_backoff_ [block]:
+  u := Time.monotonic_us
```
Comment (kasperl) [inline lib/ble.toit:653]:
```suggestion
  start := Time.monotonic_us
```

File: lib/ble.toit [hunk 6]
```diff
@@ -588,13 +626,37 @@ ble_add_server_service_ resource_group_ uuid:
   #primitive.ble.add_server_service
 
 ble_add_server_characteristic_ service_resource uuid type value:
+  ble_run_with_quota_backoff_ :
```
Comment (kasperl) [inline lib/ble.toit:629]:
```suggestion
  return ble_run_with_quota_backoff_:
    ble_add_server_characteristic_primitive_ service_resource uuid type value
```

File: lib/ble.toit [hunk 7]
```diff
@@ -588,13 +626,37 @@ ble_add_server_service_ resource_group_ uuid:
...
+    return ble_add_server_characteristic_primitive_ service_resource uuid type value
+  unreachable
+
+ble_add_server_characteristic_primitive_ service_resource uuid type value:
   #primitive.ble.add_server_characteristic
 
-ble_set_characteristics_value_ gatt new_value:
+ble_set_characteristics_value_ gatt new_value -> none:
+  ble_run_with_quota_backoff_ :
```
Comment (kasperl) [inline lib/ble.toit:637]:
```suggestion
  ble_run_with_quota_backoff_:
```

File: lib/ble.toit [hunk 8]
```diff
@@ -527,6 +561,10 @@ CONNECTED_EVENT_            ::= 1 << 3
 CONNECT_FAILED_EVENT_       ::= 1 << 4
 DISCONNECTED_EVENT_         ::= 1 << 5
 
+
```
Comment (kasperl) [inline lib/ble.toit:564]: Drop one newline.

File: lib/ble.toit [hunk 9]
```diff
@@ -428,17 +461,18 @@ If services is not empty, sets up the services for the advertiser.
 */
 class Device:
   resource_group_ := ?
-  resource_state_/monitor.ResourceState_? := null
+  resource_state_/ResourceState_? := null
 
-  constructor.default server_configuration/ServerConfiguration?=null:
+  constructor.default server_configuration/ServerConfiguration?=null --preferred_mtu/num=BLE_DEFAULT_PREFERRED_MTU_:
```
Comment (kasperl) [inline lib/ble.toit:466]:
```suggestion
  constructor.default server_configuration/ServerConfiguration?=null --preferred_mtu/int=BLE_DEFAULT_PREFERRED_MTU_:
```

File: lib/ble.toit [hunk 10]
```diff
@@ -392,16 +412,24 @@ class ReadOnlyCharacteristic extends Characteristic:
 A characteristic that can only be written to by clients.
 */
 class WriteOnlyCharacteristic extends WritableCharacteristic:
-  constructor service/Service uuid/uuid_pkg.Uuid:
-    resource := ble_add_server_characteristic_ service.resource_ uuid.to_byte_array BLE_CHR_TYPE_WRITE_ONLY_ null
+  constructor service/Service uuid/uuid_pkg.Uuid require_response/bool:
+    resource := ble_add_server_characteristic_
+        service.resource_
+        uuid.to_byte_array
+        require_response?BLE_CHR_TYPE_WRITE_ONLY_:BLE_CHR_TYPE_WRITE_ONLY_NO_RSP_
```
Comment (kasperl) [inline lib/ble.toit:419]:
```suggestion
        require_response ? BLE_CHR_TYPE_WRITE_ONLY_ : BLE_CHR_TYPE_WRITE_ONLY_NO_RSP_
```

File: lib/ble.toit [hunk 11]
```diff
@@ -366,9 +379,12 @@ abstract class WritableCharacteristic extends Characteristic:
...
   value -> ByteArray?:
-    state_.wait_for_state BLE_WAIT_RECV_
-    data := ble_get_characteristics_value_ state_.resource
-    state_.clear_state BLE_WAIT_RECV_
+    data := null
+    while not data:
+      state_.wait_for_state BLE_WAIT_RECV_
+      data = ble_get_characteristics_value_ state_.resource
+      state_.clear_state BLE_WAIT_RECV_
+
```
Comment (kasperl) [inline lib/ble.toit:387]: Drop newline?

File: lib/ble.toit [hunk 12]
```diff
@@ -358,6 +362,15 @@ abstract class Characteristic:
   constructor service/Service resource .uuid:
     state_ = ResourceState_ service.server_configuration_.resource_group_ resource
 
+  /**
+    The currently negotiated mtu of the characteristic. Only meaning full when a client is connected
```
Comment (kasperl) [inline lib/ble.toit:366]:
```suggestion
  The currently negotiated mtu of the characteristic. Only meaningful when a client is connected.
```

File: lib/ble.toit [hunk 13]
```diff
@@ -358,6 +362,15 @@ abstract class Characteristic:
...
+  */
+  att_mtu -> num:
```
Comment (kasperl) [inline lib/ble.toit:368]:
```suggestion
  att_mtu -> int:
```

File: lib/ble.toit [hunk 14]
```diff
@@ -239,7 +241,8 @@ class RemoteCharacteristic:
   Writes the value of the characteristic on the remote service.
   */
   write_value value/ByteArray -> none:
-    ble_send_data_ service.client_.gatt_ handle value
+    ble_run_with_quota_backoff_ :
```
Comment (kasperl) [inline lib/ble.toit:244]:
```suggestion
    ble_run_with_quota_backoff_:
```

File: lib/ble.toit [hunk 15]
```diff
@@ -358,6 +362,15 @@ abstract class Characteristic:
   constructor service/Service resource .uuid:
     state_ = ResourceState_ service.server_configuration_.resource_group_ resource
 
+  /**
+    The currently negotiated mtu of the characteristic. Only meaning full when a client is connected
+  */
+  att_mtu -> num:
+    low_level_response := ble_get_mtu_ state_.resource
```
Comment (kasperl) [inline lib/ble.toit:369]:
```suggestion
    mtu := ble_get_mtu_ state_.resource
```

File: src/resources/ble_esp32.cc [hunk 16]
```diff
@@ -1057,8 +1065,26 @@ PRIMITIVE(get_characteristics_value) {
   } else {
     ALLOCATION_FAILED;
   }
+}
+
+PRIMITIVE(set_preferred_mtu) {
+  ARGS(int, mtu);
+
+  int ret_val = ble_att_set_preferred_mtu(mtu);
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:1073]: We typically try to avoid abbreviations, but I know it is hard in this ble/mtu/att world :)

```suggestion
  int result = ble_att_set_preferred_mtu(mtu);
```

File: src/resources/ble_esp32.cc [hunk 17]
```diff
@@ -1057,8 +1065,26 @@ PRIMITIVE(get_characteristics_value) {
...
+    INVALID_ARGUMENT;
+  } else {
+    return process->program()->null_object();
+  }
+}
+
+PRIMITIVE(get_att_mtu) {
+  ARGS(BLEServerCharacteristicResource, resource);
+
+  word mtu = ble_att_mtu(resource->conn_handle());
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:1085]:
```suggestion
  uint16 mtu = ble_att_mtu(resource->conn_handle());
```

**Review by kasperl — CHANGES_REQUESTED** [review]
Good catch! Looking forward to seeing this go in.

What are the changes to third_party/esp-idf? They might be unrelated to this PR.

**Comment by mikkeldamsgaard** [comment]
Closing this PR, as my git skills got in the way.
